client/lib/cart-values/cart-items.ts

```diff
@@ -0,0 +1,815 @@
+ */
+export function updatePrivacyForDomain(
+	item: ResponseCartProduct,
+	value: boolean
+): ResponseCartProduct {
+	return {
+		...item,
+		extra: {
+			...item.extra,
+			privacy: value,
+		},
+	};
+}
+
+/**
+ * Determines whether a cart item is partial credits
+ */
+function isPartialCredits( cartItem: ResponseCartProduct ): boolean {
+	return cartItem.product_slug === 'wordpress-com-credits';
+}
+
+/**
+ * Returns true if, according to cart attributes, a `domain` should be free
+ */
+export function isNextDomainFree( cart?: ResponseCart, domain = '' ): boolean {
+	if ( ! cart || ! cart.next_domain_is_free ) {
+		return false;
+	}
+
+	if ( cart.next_domain_condition === 'blog' ) {
+		if ( getTld( domain ) !== 'blog' ) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+export function isDomainBundledWithPlan( cart?: ResponseCart, domain?: string ): boolean {
+	const bundledDomain = cart?.bundled_domain ?? '';
+	return '' !== bundledDomain && ( domain ?? '' ).toLowerCase() === bundledDomain.toLowerCase();
+}
+
+/**
+ * Returns true if cart contains a plan and also a domain that comes for free with that plan
+ */
+export function isDomainBeingUsedForPlan( cart?: ResponseCart, domain?: string ): boolean {
+	if ( ! cart || ! domain ) {
+		return false;
+	}
+
+	if ( ! hasPlan( cart ) ) {
+		return false;
+	}
+
+	const domainProducts = getDomainRegistrations( cart ).concat( getDomainMappings( cart ) );
+	const domainProduct = domainProducts.shift() || { meta: '' };
+	const processedDomainInCart = domain === domainProduct.meta;
+	if ( ! processedDomainInCart ) {
+		return false;
+	}
+
+	if ( hasBloggerPlan( cart ) ) {
+		const tld = domain.split( '.' ).pop();
+		if ( tld !== 'blog' ) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+function hasSomeSlug( data: unknown ): data is WithSnakeCaseSlug | WithCamelCaseSlug {
+	return Boolean(
+		( data as WithSnakeCaseSlug ).product_slug || ( data as WithCamelCaseSlug ).productSlug
+	);
+}
+
+export function shouldBundleDomainWithPlan(
+	withPlansOnly: boolean,
+	selectedSite: undefined | { plan: { product_slug: string } },
+	cart: ResponseCart,
+	suggestionOrCartItem: {
+		product_slug?: string;
+		productSlug?: string;
+		is_domain_registration?: boolean;
+		isDomainRegistration?: boolean;
+		domain_name?: string;
+		is_free?: boolean;
+	}
+): boolean {
+	return Boolean(
+		withPlansOnly &&
+			// not free or a cart item
+			( isDomainRegistration( suggestionOrCartItem ) ||
+				( hasSomeSlug( suggestionOrCartItem ) && isDomainMapping( suggestionOrCartItem ) ) ||
+				( suggestionOrCartItem.domain_name &&
+					! isFreeWordPressComDomain( suggestionOrCartItem ) ) ) &&
+			! isDomainBeingUsedForPlan( cart, suggestionOrCartItem.domain_name ) && // a plan in cart
+			! isNextDomainFree( cart ) && // domain credit
+			! hasPlan( cart ) && // already a plan in cart
+			( ! selectedSite || ( selectedSite && selectedSite.plan.product_slug === 'free_plan' ) )
```

Comment:
Nit: since we're checking for `! selectedSite` in the first part of the condition, this means that in the second one it's not falsy. That means that the `selectedSite && ` check is redundant. 

I realize this was here before, so feel free to ignore my cleanup suggestion.